### ISSUE TITLE
Custom layer parameters aren't counted in model summary

### ISSUE THREAD

**Issue by jonathanbratt** [rstudio/keras#926]
When I construct a model using a custom layer, the parameters in the custom layer aren't counted in the model summary. For example, here's a simple model consisting of three dense layers:

```
library(keras)
model_custom <- keras_model_sequential()
model_custom %>% 
    layer_dense(units = 256L, input_shape = c(784L)) %>%
    custom_layer_dense(units = 256L) %>%
    layer_dense(units = 10L, activation = 'softmax')
```

...where `custom_layer_dense` is just a wrapper around a standard dense layer:

```
CustomDense <- R6::R6Class(
    "CustomDense",
    inherit = KerasLayer,
    lock_objects = FALSE,
    
    public = list(
        initialize = function(units) {
            self$units <- units
        },
        
        build = function(input_shape) {
            self$dense <- keras::layer_dense(units = self$units)
            super$build(input_shape)
        },
        
        call = function(input, mask = NULL) {
            output <- self$dense(input) 
        },
        
        compute_output_shape = function(input_shape) {
            input_shape[[2]] <- self$units 
            input_shape
        }
    )
)

custom_layer_dense <- function(object, 
                               units,
                               name = NULL, 
                               trainable = TRUE) {
    create_layer(CustomDense, 
                 object, 
                 args = list(units = units,
                             name = name,
                             trainable = trainable))
}
```
When training, this model appears to behave the same as a model using a standard dense layer. However, calling the `summary` method produces:

```
> model_custom$summary()
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_1 (Dense)              (None, 256)               200960    
_________________________________________________________________
r_layer (RLayer)             (None, 256)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                2570      
=================================================================
Total params: 203,530
Trainable params: 203,530
Non-trainable params: 0
_________________________________________________________________
```
The middle dense layer is attributed no parameters. Also, `model_custom$weights` yields only the weights for the built-in layers, not the custom layer.
Am I doing something wrong in my layer construction, or is this due to limitations in the`RLayer` class?

**Comment by dfalbel** [comment]
This looks like a bug to me. Note that the following works in python:

```
import tensorflow as tf

class CustomLayer(tf.keras.layers.Layer):
  
  def __init__(self, units, **kwargs):
    self.units = units
    super(CustomLayer, self).__init__(**kwargs)
    
  def build (self, input_shape):
    self.dense = tf.keras.layers.Dense(units = self.units)
    super(CustomLayer, self).build(input_shape)
    
  def call(self, x):
    return self.dense(x)
    
  def compute_output_shape(self, input_shape):
    (input_shape[0], self.units)
    

model = tf.keras.Sequential([
  CustomLayer(units = 10, input_shape = (5,)),
  tf.keras.layers.Dense(units = 5)
])

model.summary()
```

**Comment by dfalbel** [comment]
I suggest the following workaround for now:

```
CustomDense <- reticulate::PyClass(
  "CustomDense",
  inherit = tensorflow::tf$keras$layers$Layer,
  list(
    
    `__init__` = function(self, units, ...) {
      self$units <- units
      super()$`__init__`()
    },
    
    build = function(self, input_shape) {
      self$dense <- layer_dense(units = reticulate::py_to_r(self$units))
      super()$build(input_shape)
    },
    
    call = function(self, x, ...) {
      self$dense(x)
    },
    
    compute_output_shape = function(input_shape) {
      list(input_shape[[1]], self$units)
    }
    
  )
)
```

This required the dev version of reticulate.

**Comment by jonathanbratt** [comment]
Excellent, that works.
Thank you very much!

**Comment by jonathanbratt** [comment]
To help me prioritize my work: do you have an estimate for how long it might take to get a bug fix into the dev version?

**Comment by dfalbel** [comment]
@jonathanbratt I am considering making the `PyClass` approach the recommended in Keras. Still need to do more testing, but it's theoretically much more general than the actual solution (eg. you can subclass a Dense layer if you want or, a Callback with the same interface.)  What do you think?

**Comment by jonathanbratt** [comment]
I could live with that. The main downside that I see is the need to `reticulate::py_to_r` every time we want to pass a class variable to an R function.

Also, I had just learned how to document R6 classes with roxygen. :) How would you suggest handling custom layers inside packages, if you go the `PyClass` route?

**Comment by dfalbel** [comment]
> The main downside that I see is the need to reticulate::py_to_r every time we want to pass a class variable to an R function. 

I'm pretty sure this is fixable in `reticulate` side. I'll take a look next week.

The workflow would be similar to the current one. Build the private class and export a function that wraps it. For example:

```CustomDense <- reticulate::PyClass(
  "CustomDense",
  inherit = tensorflow::tf$keras$layers$Layer,
  list(
    
    `__init__` = function(self, units, ...) {
      self$units <- units
      super()$`__init__`()
    },
    
    build = function(self, input_shape) {
      self$dense <- layer_dense(units = reticulate::py_to_r(self$units))
      super()$build(input_shape)
    },
    
    call = function(self, x, ...) {
      self$dense(x)
    },
    
    compute_output_shape = function(input_shape) {
      list(input_shape[[1]], self$units)
    }
    
  )
)

#' @export
layer_custom_dense(object, units, name = NULL, trainable = NULL, weights = NULL) {
  create_layer(CustomDense, object, list(
   units = units, 
   name = name,
   trainable = trainable,
   weights = weights
  ))
}```

**Comment by jonathanbratt** [comment]
In case it would be helpful to you to see examples of actual use cases, I have several R6 custom layers in an RBERT dev branch here: 
https://github.com/jonathanbratt/RBERT/tree/tf2-dev/R
(See files starting with 'tf2-layer.')

**Comment by quocdat32461997** [comment]
When I create the Multi-Head-Attention layer below similar to the following link [Tensorflow Transformer](https://www.tensorflow.org/tutorials/text/transformer#masking) 
`class MultiHeadAttention(keras.layers.Layer):
    def __init__(self, d_model, num_heads, trainable = False, name = 'MHA', **kwargs):
        self.num_heads = num_heads
        self.d_model = d_model
        self.trainable = trainable
        self.name = name
    
        assert d_model % self.num_heads == 0
        self.depth = self.d_model // self.num_heads
        super(MultiHeadAttention, self).__init__(**kwargs)
        
    def build(self, input_shape):
        
        self.wq = Dense(self.d_model, activation = 'relu', trainable = self.trainable)
        self.wk = Dense(self.d_model, activation = 'relu', trainable = self.trainable)
        self.wv = Dense(self.d_model, activation = 'relu', trainable = self.trainable)
    
        self.dense = keras.layers.Dense(self.d_model, trainable = self.trainable)
        
        super(MultiHeadAttention, self).build(input_shape)
    
    def _scaled_dot_product_attention(self, q, k, v, mask = None):
        """Calculate the attention weights.
        q, k, v must have matching leading dimensions.
        k, v must have matching penultimate dimension, i.e.: seq_len_k = seq_len_v.
        The mask has different shapes depending on its type(padding or look ahead) 
        but it must be broadcastable for addition.

            Args:
                q: query shape == (..., seq_len_q, depth)
                k: key shape == (..., seq_len_k, depth)
                v: value shape == (..., seq_len_v, depth_v)
                mask: Float tensor with shape broadcastable to (..., seq_len_q, seq_len_k). Defaults to None.

            Returns:
                output, attention_weights
            """
        matmul_qk = tf.matmul(q, k, transpose_b=True)  # (..., seq_len_q, seq_len_k)
      
        #matmul_qk = K.dot(q, k)
        # scale matmul_qk
        dk = tf.cast(tf.shape(k)[-1], tf.float32)
        scaled_attention_logits = matmul_qk / tf.math.sqrt(dk)
        
        # add the mask to the scaled tensor.
        if mask is not None:
            scaled_attention_logits += (mask * -1e9)  

        # softmax is normalized on the last axis (seq_len_k) so that the scores
        # add up to 1.
        attention_weights = tf.nn.softmax(scaled_attention_logits, axis=-1)  # (..., seq_len_q, seq_len_k)
        output = tf.matmul(attention_weights, v)  # (..., seq_len_q, depth_v)
        return output, attention_weights
        
    def split_heads(self, x, batch_size):
        """Split the last dimension into (num_heads, depth).
        Transpose the result such that the shape is (batch_size, num_heads, seq_len, depth)
        """
        #x = tf.reshape(x, (batch_size, -1, self.num_heads, self.depth))
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.depth))
        return tf.transpose(x, perm=[0, 2, 1, 3])
    """
    def build(self, input_shape):
        #Create trainable weight varaible for the Multi-Head-Attention Layer
        self.kernel = self.add_weight(name = 'kernel', shape = (input_shape[0][1], self.output_dim), \
                                     initialier = 'uniform', trainable = self.trainable)
        super(MultiHeadAttention, self).build(input_shape)
    """
    
    def call(self, inputs, mask = None):
        batch_size = tf.shape(inputs)[0]
    
        q = self.wq(inputs)  # (batch_size, seq_len, d_model)
        k = self.wk(inputs)  # (batch_size, seq_len, d_model)
        v = self.wv(inputs)  # (batch_size, seq_len, d_model)
    
        q = self.split_heads(q, batch_size)  # (batch_size, num_heads, seq_len_q, depth)
        k = self.split_heads(k, batch_size)  # (batch_size, num_heads, seq_len_k, depth)
        v = self.split_heads(v, batch_size)  # (batch_size, num_heads, seq_len_v, depth)
    
        # scaled_attention.shape == (batch_size, num_heads, seq_len_q, depth)
        # attention_weights.shape == (batch_size, num_heads, seq_len_q, seq_len_k)
        scaled_attention, attention_weights = self._scaled_dot_product_attention( q, k, v, mask)
    
        scaled_attention = tf.transpose(scaled_attention, perm=[0, 2, 1, 3])  # (batch_size, seq_len_q, num_heads, depth)

        concat_attention = tf.reshape(scaled_attention, (batch_size, -1, self.d_model))  # (batch_size, seq_len_q, d_model)

        output = self.dense(concat_attention)  # (batch_size, seq_len_q, d_model)
   
        return output
`
I got the no parameter for the Multi-Head-Attention-Layer when adding the custom layer by either Sequential or stacking with Input layer from pure Keras. I am using Tensorflow 2.1.0 and Keras 2.3.1 on AMD Radeon VII
![Screenshot from 2020-04-06 02-35-44](https://user-images.githubusercontent.com/18373768/78533945-5b1f9500-77af-11ea-9878-610ed41385b2.png)

**Comment by willi3by** [comment]
@dfalbel is there now a method implemented to solve this issue with R6 custom layers? I apologize if I am missing something. 

Thanks!

**Comment by jonathanbratt** [comment]
Hi @willi3by.
This issue is not solved in R6 custom layers, but in the dev version of keras there is a new function (keras::Layer) for creating custom layers which does solve this problem. I think there is also a new vignette with examples of how to use the new Layer function.

**Comment by willi3by** [comment]
Thank you for the information!

**Comment by anugrahasinha** [comment]
Although this if for R, however, python also has the issue quite similar to this.
Ref : https://github.com/tensorflow/tensorflow/issues/28748
